c_glib/gandiva-glib/function-registry.cpp

```diff
@@ -0,0 +1,179 @@
+{
+  auto priv = GGANDIVA_FUNCTION_REGISTRY_GET_PRIVATE(object);
+
+  if (priv->native_function_list) {
+    GList *p = priv->native_function_list;
+    while (p) {
+      g_object_unref(G_OBJECT(p->data));
+      p = g_list_next(p);
+    }
+    g_list_free(priv->native_function_list);
+  }
+
+  delete priv->function_registry;
+
+  G_OBJECT_CLASS(ggandiva_function_registry_parent_class)->finalize(object);
+}
+
+static void
+ggandiva_function_registry_set_property(GObject *object,
+                                        guint prop_id,
+                                        const GValue *value,
+                                        GParamSpec *pspec)
+{
+  auto priv = GGANDIVA_FUNCTION_REGISTRY_GET_PRIVATE(object);
+  switch (prop_id) {
+  case PROP_FUNCTION_REGISTRY:
+    {
+      auto ptr = g_value_get_pointer(value);
+      auto gandiva_function_registry
+        = static_cast<std::shared_ptr<gandiva::FunctionRegistry> *>(ptr);
+      priv->function_registry
+        = new std::shared_ptr<gandiva::FunctionRegistry>(*gandiva_function_registry);
+    }
+    break;
+  default:
+    G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+    break;
+  }
+}
+
+static void
+ggandiva_function_registry_init(GGandivaFunctionRegistry *object)
+{
+}
+
+static void
+ggandiva_function_registry_class_init(GGandivaFunctionRegistryClass *klass)
+{
+  auto gobject_class = G_OBJECT_CLASS(klass);
+
+  gobject_class->dispose = ggandiva_function_registry_dispose;
+  gobject_class->finalize = ggandiva_function_registry_finalize;
+  gobject_class->set_property = ggandiva_function_registry_set_property;
+
+  GParamSpec *spec;
+  spec = g_param_spec_pointer("function_registry",
+                              "FunctionRegistry",
+                              "The raw gandiva::FunctionRegistry *",
+                              static_cast<GParamFlags>(G_PARAM_WRITABLE |
+                                                       G_PARAM_CONSTRUCT_ONLY));
+  g_object_class_install_property(gobject_class, PROP_FUNCTION_REGISTRY, spec);
+}
+
+/**
+ * ggandiva_function_registry_new:
+ *
+ * Returns: A newly created #GGandivaFunctionRegistry.
+ */
+GGandivaFunctionRegistry *
+ggandiva_function_registry_new(void)
+{
+  auto gandiva_function_registry
+    = std::make_shared<gandiva::FunctionRegistry>();
+  return ggandiva_function_registry_new_raw(&gandiva_function_registry);
+}
+
+/**
+ * ggandiva_function_registry_get_native_functions:
+ * @function_registry: A #GGandivaFunctionRegistry.
+ *
+ * Returns: (transfer none) (element-type GGandivaNativeFunction):
+ *   The native functions in the function registry.
+ */
+GList *
+ggandiva_function_registry_get_native_functions(GGandivaFunctionRegistry *function_registry)
+{
+  auto priv = GGANDIVA_FUNCTION_REGISTRY_GET_PRIVATE(function_registry);
+
+  if (!priv->native_function_list) {
+    using iterator = gandiva::FunctionRegistry::iterator;
+
+    const auto& gandiva_function_registry = *priv->function_registry;
```

Comment:
Could you use C like `const auto &g...` style for `&` (and `*`)?
This is C++ code but we use C like style because this is GLib style code.

Reply:
Anyway, this should be:

```c++
gandiva::FunctionRegistry registry;
```